File: Leetcode/0057 Insert Interval.cpp
class Solution {
    void merge(vector<vector<int>>& intervals, const vector<int>& new_interval) {
        if(intervals.empty()) {
            intervals.push_back(new_interval);
            return;
        }
        auto &last_interval = intervals.back();
        if(last_interval[1] < new_interval[0]) {
            intervals.push_back(new_interval);
        } else {
            last_interval[1] = max(last_interval[1], new_interval[1]);
        }
    }
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> merged_intervals;
        bool inserted = false;
        for(int i = 0;i <= intervals.size();) {
            bool is_array_over = (i == intervals.size());
            bool can_insert_new_interval = (not inserted) and (is_array_over or newInterval[0] < intervals[i][0]);
            if(can_insert_new_interval) {
                merge(merged_intervals, newInterval);
                inserted = true;
            } else {
                if(!is_array_over) {
                    merge(merged_intervals, intervals[i]);                    
                }
                ++i;                
            }
        }
        return merged_intervals;
    }
};

File: Leetcode/0057 Insert Interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int positionOfNewInterval = intervals.length;
        for(int i = 0;i < intervals.length; ++i) {
            if(newInterval[0] < intervals[i][0]) {
                positionOfNewInterval = i;
                break;
            }
        }

        List<int[]> afterInsert = new ArrayList<>();
        for(int i = 0;i < positionOfNewInterval; ++i) {
            afterInsert.add(intervals[i]);
        }

        insertInverval(afterInsert, newInterval);
        
        for(int i = positionOfNewInterval;i < intervals.length; ++i) {
            insertInverval(afterInsert, intervals[i]);
        }
        return afterInsert.toArray(new int[0][]);
    }

    private boolean isNotOverlapping(int[] a,int[] b) {
        return a[1] < b[0] || b[1] < a[0];
    }

    private void insertInverval(List<int[]> sortedIntervals, int[] interval) {
        if(sortedIntervals.isEmpty()) {
            sortedIntervals.add(interval);
            return;
        }
        var lastInterval = sortedIntervals.get(sortedIntervals.size() - 1);
        if(isNotOverlapping(interval, lastInterval)) {
            sortedIntervals.add(interval);
            return;
        }
        lastInterval[1] = Math.max(lastInterval[1], interval[1]); 
    }
}
